File: scripts/nsfw_scan.py
#!/usr/bin/env python3
"""
scripts/nsfw_scan.py
────────────────────────────────────────────────────────────────────────────
NudeNet-based NSFW scanner for the Exhibit archive.

NudeNet uses body-part detection (not scene classification), which means
it correctly identifies nudity in classical paintings, engravings, drawings,
and photographs — not just modern photos.

Install:
    pip install nudenet pillow

Usage:
    python scripts/nsfw_scan.py               # scan everything
    python scripts/nsfw_scan.py --limit 5000  # test on first 5000 items
    python scripts/nsfw_scan.py --threshold 0.5  # stricter (fewer false positives)

Output:
    scripts/nsfw-flagged.json  — review this, remove false positives, then run:
    node scripts/nsfw-purge.mjs
────────────────────────────────────────────────────────────────────────────
"""

import sqlite3
import hashlib
import os
import sys
import json
import time
import argparse
import re
from pathlib import Path

# ── Config ────────────────────────────────────────────────────────────────────
SCRIPT_DIR = Path(__file__).parent
DB_PATH    = SCRIPT_DIR.parent / "artworks.db"
CACHE_DIR  = Path(r"C:\Users\ameen\Desktop\.img-cache")
OUTPUT     = SCRIPT_DIR / "nsfw-flagged.json"

# NudeNet confidence threshold — 0.4 is a good balance.
# Lower = more flags (more false positives).
# Higher = fewer flags (might miss some).
DEFAULT_THRESHOLD = 0.4

# These NudeNet labels are considered NSFW
NSFW_LABELS = {
    "FEMALE_GENITALIA_EXPOSED",
    "MALE_GENITALIA_EXPOSED",
    "FEMALE_BREAST_EXPOSED",
    "ANUS_EXPOSED",
    "BUTTOCKS_EXPOSED",
}

# These are detected but less severe — flagged at lower confidence
PARTIAL_LABELS = {
    "FEMALE_BREAST_COVERED",
    "FEMALE_GENITALIA_COVERED",
    "MALE_GENITALIA_COVERED",
}

# Strong metadata keywords — flagged regardless of model
STRONG_KEYWORDS = [
    "nude", "nudity", "naked", "nakedness",
    "erotic", "erotica", "pornographic", "obscene",
]

# ── Cache filename logic — exact copy of getFetchUrl from prewarm-cache.mjs ──
def iiif_url(url: str, size: int) -> str:
    return re.sub(r"/full/[^/]+/", f"/full/!{size},{size}/", url)

def ch_source_url(url: str) -> str:
    return re.sub(r"_[bzn]\.jpg$", "_z.jpg", url, flags=re.IGNORECASE)

def get_fetch_url(image_url: str, size: int) -> str:
    if "letterformarchive.org" in image_url:
        return image_url
    if "/full/" in image_url:
        return iiif_url(image_url, size)
    if "images.collection.cooperhewitt.org" in image_url and \
       re.search(r"_[bzn]\.jpg$", image_url, re.IGNORECASE):
        return ch_source_url(image_url)
    return image_url

def get_cache_filename(image_url: str, size: int) -> str:
    fetch_url = get_fetch_url(image_url, size)
    key       = f"{fetch_url}:{size}"
    return hashlib.md5(key.encode()).hexdigest() + ".jpg"

def find_cached_path(image_url: str) -> Path | None:
    for size in [400, 1200]:
        p = CACHE_DIR / get_cache_filename(image_url, size)
        try:
            if p.stat().st_size > 500:
                return p
        except FileNotFoundError:
            pass
    return None

def get_keyword_match(row: dict) -> str | None:
    haystack = " ".join(filter(None, [
        row.get("title"), row.get("classification"),
        row.get("medium"), row.get("objectType"), row.get("subCategory"),
    ])).lower()
    return next((kw for kw in STRONG_KEYWORDS if kw in haystack), None)

# ── Main ──────────────────────────────────────────────────────────────────────
def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--limit",     type=int,   default=None, help="Only scan first N items (for testing)")
    parser.add_argument("--threshold", type=float, default=DEFAULT_THRESHOLD, help="Detection confidence threshold (default 0.4)")
    parser.add_argument("--category",  type=str,   default=None, help="Only scan a specific category e.g. 'Painting'")
    args = parser.parse_args()

    print("\n🔍 NSFW Scanner (NudeNet) — Exhibit Archive")
    print("─────────────────────────────────────────────\n")

    # Check DB
    if not DB_PATH.exists():
        print(f"❌ artworks.db not found at {DB_PATH}")
        sys.exit(1)

    # Check cache dir
    if not CACHE_DIR.exists():
        print(f"❌ Cache directory not found: {CACHE_DIR}")
        print("   Update CACHE_DIR at the top of this script.")
        sys.exit(1)

    # Load NudeNet
    print("📦 Loading NudeNet model...")
    print("   (First run downloads ~90MB model — subsequent runs are instant)\n")
    try:
        from nudenet import NudeDetector
        detector = NudeDetector()
    except ImportError:
        print("❌ NudeNet not installed. Run: pip install nudenet pillow")
        sys.exit(1)
    print("✅ Model ready\n")

    # Load artworks from DB
    conn = sqlite3.connect(str(DB_PATH))
    conn.row_factory = sqlite3.Row
    cur  = conn.cursor()
    if args.category:
        cur.execute("""
            SELECT id, title, author, year, imageUrl, source, mainCategory, link,
                   classification, medium, objectType, subCategory
            FROM artworks
            WHERE imageUrl IS NOT NULL AND imageUrl != ''
              AND mainCategory = ?
        """, (args.category,))
    else:
        cur.execute("""
            SELECT id, title, author, year, imageUrl, source, mainCategory, link,
                   classification, medium, objectType, subCategory
            FROM artworks
            WHERE imageUrl IS NOT NULL AND imageUrl != ''
        """)
    rows = [dict(r) for r in cur.fetchall()]
    conn.close()

    if args.limit:
        rows = rows[:args.limit]
        print(f"⚠️  Limited to first {args.limit} items\n")

    # Spot-check cache
    found = sum(1 for r in rows[:200] if find_cached_path(r["imageUrl"]))
    print(f"📊 {len(rows):,} artworks to scan")
    print(f"   Cache spot-check (first 200): {found}/200 found")
    if found < 50:
        print(f"\n❌ Too few cached files found. Check CACHE_DIR: {CACHE_DIR}")
        sys.exit(1)
    print(f"   Threshold: {args.threshold}\n")

    flagged   = []
    cache_miss = 0
    errors    = 0
    scanned   = 0
    start     = time.time()

    for row in rows:
        scanned += 1
        img_path = find_cached_path(row["imageUrl"])
        kw_match = get_keyword_match(row)

        # Strong keyword match — flag even without image
        if kw_match and img_path is None:
            flagged.append({
                "id":           row["id"],
                "title":        row["title"],
                "author":       row["author"],
                "year":         row["year"],
                "source":       row["source"],
                "category":     row["mainCategory"],
                "link":         row["link"],
                "imageUrl":     row["imageUrl"],
                "keywordMatch": kw_match,
                "detections":   None,
                "triggeredBy":  [f"keyword:{kw_match}"],
            })
            cache_miss += 1
            continue

        if img_path is None:
            cache_miss += 1
            continue

        try:
            detections = detector.detect(str(img_path))

            # Filter to NSFW labels above threshold
            nsfw_hits    = [d for d in detections if d["class"] in NSFW_LABELS    and d["score"] >= args.threshold]
            partial_hits = [d for d in detections if d["class"] in PARTIAL_LABELS and d["score"] >= args.threshold + 0.2]
            all_hits     = nsfw_hits + partial_hits

            # Also flag on strong keyword even if model is clean
            should_flag = len(all_hits) > 0 or (kw_match is not None)

            if should_flag:
                flagged.append({
                    "id":          row["id"],
                    "title":       row["title"],
                    "author":      row["author"],
                    "year":        row["year"],
                    "source":      row["source"],
                    "category":    row["mainCategory"],
                    "link":        row["link"],
                    "imageUrl":    row["imageUrl"],
                    "keywordMatch": kw_match,
                    "detections":  [{"label": d["class"], "score": round(d["score"], 3)} for d in all_hits],
                    "triggeredBy": list({d["class"] for d in all_hits}) + ([f"keyword:{kw_match}"] if kw_match else []),
                })

        except Exception as e:
            errors += 1

        # Progress
        if scanned % 100 == 0 or scanned == len(rows):
            elapsed  = time.time() - start
            rate     = scanned / elapsed if elapsed > 0 else 1
            rem_secs = (len(rows) - scanned) / rate
            rem_h    = rem_secs / 3600
            pct      = (scanned / len(rows)) * 100
            print(
                f"\r  {pct:5.1f}% ({scanned:,}/{len(rows):,}) | "
                f"flagged: {len(flagged)} | "
                f"{elapsed:.0f}s | ~{rem_h:.1f}h left   ",
                end="", flush=True
            )

        # Save checkpoint every 5,000
        if scanned % 5000 == 0:
            _write_output(flagged, scanned, len(rows), cache_miss, errors, "in-progress", args.threshold)

    print("\n")

    _write_output(flagged, scanned, len(rows), cache_miss, errors, "complete", args.threshold)

    total_min = (time.time() - start) / 60
    print(f"✅ Done in {total_min:.1f} minutes")
    print(f"   Scanned:   {scanned:,}")
    print(f"   Flagged:   {len(flagged):,}")
    print(f"   No cache:  {cache_miss:,}")
    print(f"   Errors:    {errors:,}")
    print(f"\n📄 Review: scripts/nsfw-flagged.json")
    print(f"   Remove any false positives, then run:")
    print(f"   node scripts/nsfw-purge.mjs\n")


def _write_output(flagged, scanned, total, cache_miss, errors, status, threshold):
    import datetime
    output = {
        "scannedAt":    datetime.datetime.now().isoformat(),
        "status":       status,
        "totalScanned": scanned,
        "totalItems":   total,
        "cacheMiss":    cache_miss,
        "errors":       errors,
        "flaggedCount": len(flagged),
        "threshold":    threshold,
        "flagged":      flagged,
    }
    with open(OUTPUT, "w", encoding="utf-8") as f:
        json.dump(output, f, indent=2, ensure_ascii=False)


if __name__ == "__main__":
    main()

File: scripts/nsfw-scan.mjs
// scripts/nsfw-scan.mjs
import Database from 'better-sqlite3';
import crypto   from 'crypto';
import fs       from 'fs';
import path     from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const DB_PATH   = path.join(__dirname, '..', 'artworks.db');
const CACHE_DIR = 'C:\\Users\\ameen\\Desktop\\.img-cache';
const OUTPUT    = path.join(__dirname, 'nsfw-flagged.json');
const BATCH_SIZE = 32;

const THRESHOLDS = { Porn: 0.40, Sexy: 0.60, Hentai: 0.50 };

const STRONG_KEYWORDS = ['nude','nudity','naked','nakedness','erotic','erotica','pornographic','obscene'];
const ALL_KEYWORDS    = [...STRONG_KEYWORDS,
  'venus','diana','danae','leda','susanna','bathsheba','nymph','nymphs',
  'satyr','satyrs','bacchante','figure study','life study','life drawing',
  'academy figure','academie','bather','bathers','bathing',
  'reclining figure','reclining woman','reclining man','odalisque','harem','torso',
];

// ── EXACT copies of getFetchUrl + getCacheFilename from prewarm-cache.mjs ────
function iiifUrl(url, size) {
  return url.replace(/\/full\/[^/]+\//, `/full/!${size},${size}/`);
}
function chSourceUrl(url) {
  return url.replace(/_[bzn]\.jpg$/i, '_z.jpg');
}
function getFetchUrl(imageUrl, size) {
  if (imageUrl.includes('letterformarchive.org')) return imageUrl;
  if (imageUrl.includes('/full/')) return iiifUrl(imageUrl, size);
  if (imageUrl.includes('images.collection.cooperhewitt.org') && /_[bzn]\.jpg$/i.test(imageUrl)) {
    return chSourceUrl(imageUrl);
  }
  return imageUrl;
}
function getCacheFilename(imageUrl, size) {
  const fetchUrl = getFetchUrl(imageUrl, size);
  const key      = `${fetchUrl}:${size}`;
  return `${crypto.createHash('md5').update(key).digest('hex')}.jpg`;
}
function findCachedPath(imageUrl) {
  for (const size of [400, 1200]) {
    const p = path.join(CACHE_DIR, getCacheFilename(imageUrl, size));
    try { if (fs.statSync(p).size > 500) return p; } catch {}
  }
  return null;
}
// ─────────────────────────────────────────────────────────────────────────────

function getKeywordMatch(row) {
  const haystack = [row.title, row.classification, row.medium, row.objectType, row.subCategory]
    .filter(Boolean).join(' ').toLowerCase();
  return ALL_KEYWORDS.find(kw => haystack.includes(kw)) || null;
}

async function main() {
  console.log('\n🔍 NSFW Full Batch Scanner — Exhibit Archive');
  console.log('─────────────────────────────────────────────\n');

  if (!fs.existsSync(DB_PATH)) { console.error('❌ artworks.db not found'); process.exit(1); }

  const db   = new Database(DB_PATH, { readonly: true });
  const rows = db.prepare(`
    SELECT id, title, author, year, imageUrl, source, mainCategory, link,
           classification, medium, objectType, subCategory
    FROM artworks WHERE imageUrl IS NOT NULL AND imageUrl != ''
  `).all();
  db.close();

  // Quick cache check to verify fix worked
  let cacheCheck = 0;
  for (const row of rows.slice(0, 200)) {
    if (findCachedPath(row.imageUrl)) cacheCheck++;
  }
  console.log(`📊 ${rows.length.toLocaleString()} artworks to scan`);
  console.log(`   Cache spot-check (first 200): ${cacheCheck}/200 found`);
  if (cacheCheck < 50) {
    console.error('\n❌ Still finding very few cached files — CACHE_DIR path may be wrong.');
    console.error(`   Currently looking in: ${CACHE_DIR}`);
    console.error('   Update CACHE_DIR at the top of this script to match your actual path.\n');
    process.exit(1);
  }
  console.log(`   Batch size: ${BATCH_SIZE}\n`);

  console.log('📦 Loading TensorFlow...');
  const tf = await import('@tensorflow/tfjs');
  await import('@tensorflow/tfjs-backend-cpu');
  await tf.setBackend('cpu');
  await tf.ready();

  console.log('📦 Loading nsfwjs model...');
  const nsfwjs = await import('nsfwjs');
  const model  = await nsfwjs.load('MobileNetV2Mid', { size: 224 });
  const sharp  = (await import('sharp')).default;
  console.log('✅ Ready\n');

  const flagged   = [];
  let scanned     = 0;
  let cacheMiss   = 0;
  let errors      = 0;
  const start     = Date.now();

  for (let i = 0; i < rows.length; i += BATCH_SIZE) {
    const batch   = rows.slice(i, i + BATCH_SIZE);
    const tensors = [];
    const valid   = [];

    for (const row of batch) {
      const imgPath = findCachedPath(row.imageUrl);

      if (!imgPath) {
        cacheMiss++;
        // Strong keyword match — flag even without image
        const kw = getKeywordMatch(row);
        if (kw && STRONG_KEYWORDS.includes(kw)) {
          flagged.push({
            id: row.id, title: row.title, author: row.author, year: row.year,
            source: row.source, category: row.mainCategory, link: row.link,
            imageUrl: row.imageUrl, keywordMatch: kw, scores: null,
            triggeredBy: [`keyword:${kw}`],
          });
        }
        continue;
      }

      try {
        const { data, info } = await sharp(imgPath)
          .resize(224, 224, { fit: 'fill' })
          .removeAlpha()
          .raw()
          .toBuffer({ resolveWithObject: true });

        tensors.push(tf.tensor3d(new Float32Array(data), [info.height, info.width, info.channels]).div(255.0));
        valid.push(row);
      } catch { errors++; }
    }

    if (tensors.length > 0) {
      try {
        const batchTensor = tf.stack(tensors);
        const predictions = await model.classify(batchTensor, tensors.length);
        batchTensor.dispose();
        tensors.forEach(t => t.dispose());

        for (let j = 0; j < valid.length; j++) {
          const row   = valid[j];
          const preds = Array.isArray(predictions[0]) ? predictions[j] : predictions;
          const scores    = Object.fromEntries(preds.map(p => [p.className, p.probability]));
          const triggered = Object.entries(THRESHOLDS).filter(([cls, thr]) => (scores[cls] || 0) >= thr);
          const kw        = getKeywordMatch(row);
          const strongKw  = kw && STRONG_KEYWORDS.includes(kw);

          if (triggered.length > 0 || strongKw) {
            flagged.push({
              id: row.id, title: row.title, author: row.author, year: row.year,
              source: row.source, category: row.mainCategory, link: row.link,
              imageUrl: row.imageUrl, keywordMatch: kw || null,
              scores: {
                Neutral: +(scores.Neutral||0).toFixed(3),
                Drawing: +(scores.Drawing||0).toFixed(3),
                Sexy:    +(scores.Sexy   ||0).toFixed(3),
                Porn:    +(scores.Porn   ||0).toFixed(3),
                Hentai:  +(scores.Hentai ||0).toFixed(3),
              },
              triggeredBy: [
                ...triggered.map(([cls]) => cls),
                ...(kw ? [`keyword:${kw}`] : []),
              ],
            });
          }
        }
      } catch (err) {
        tensors.forEach(t => { try { t.dispose(); } catch {} });
        errors++;
      }
    }

    scanned += batch.length;

    // Save checkpoint every 5k items
    if (scanned % 5000 < BATCH_SIZE) {
      fs.writeFileSync(OUTPUT, JSON.stringify({
        status: 'in-progress', progress: `${scanned}/${rows.length}`,
        flaggedCount: flagged.length, thresholds: THRESHOLDS, flagged,
      }, null, 2));
    }

    const elapsed  = ((Date.now() - start) / 1000).toFixed(0);
    const pct      = ((scanned / rows.length) * 100).toFixed(1);
    const rate     = scanned / ((Date.now() - start) / 1000);
    const remH     = ((rows.length - scanned) / rate / 3600).toFixed(1);
    process.stdout.write(`\r  ${pct}% (${scanned.toLocaleString()}/${rows.length.toLocaleString()}) | flagged: ${flagged.length} | ${elapsed}s | ~${remH}h left   `);
  }

  console.log('\n');

  fs.writeFileSync(OUTPUT, JSON.stringify({
    scannedAt: new Date().toISOString(), status: 'complete',
    totalScanned: scanned, cacheMiss, errors,
    flaggedCount: flagged.length, thresholds: THRESHOLDS, flagged,
  }, null, 2));

  const totalMin = ((Date.now() - start) / 60000).toFixed(1);
  console.log(`✅ Done in ${totalMin} minutes`);
  console.log(`   Scanned:  ${scanned.toLocaleString()}`);
  console.log(`   Flagged:  ${flagged.length.toLocaleString()}`);
  console.log(`   No cache: ${cacheMiss.toLocaleString()}`);
  console.log(`   Errors:   ${errors}`);
  console.log(`\n📄 Review: scripts/nsfw-flagged.json`);
  console.log(`   Then run: node scripts/nsfw-purge.mjs\n`);
}

main().catch(err => { console.error('\n❌ Fatal:', err.message); process.exit(1); });
